compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseConsumerIntoLoop.cpp

```diff
@@ -49,72 +46,74 @@ void AMDAIEFuseConsumerIntoLoopPass::runOnOperation() {
-          }
-          return WalkResult::advance();
-        });
-
-    if (!scfLoopOp) {
-      LLVM_DEBUG(llvm::dbgs()
-                 << "There is no scf.for/forall loop to fuse with\n");
-      return;
-    }
-
-    // Search the compute op and its consumer slices.
-    linalg::LinalgOp linalgOp;
-    scfLoopOp->walk<WalkOrder::PostOrder, ReverseIterator>(
-        [&](linalg::LinalgOp op) {
-          linalgOp = op;
+  // Step 2. Find the first scf loop in postorder walk.
+  Operation *scfLoopOp = nullptr;
+  funcOp->walk<WalkOrder::PostOrder, ReverseIterator>(
+      [&](LoopLikeOpInterface op) {
+        if (isa<scf::ForOp>(op)) {
+          scfLoopOp = op;
           return WalkResult::interrupt();
-        });
-
-    if (!linalgOp) {
-      LLVM_DEBUG(llvm::dbgs() << "Could not find any compute op\n");
-      return;
-    }
-
-    Value::user_range users = linalgOp->getResult(0).getUsers();
-    if (!llvm::hasSingleElement(users)) {
-      linalgOp->emitOpError("Expected only one user of the compute op");
-      return signalPassFailure();
-    }
+        } else if (isa<scf::ForallOp>(op)) {
+          scfLoopOp = op;
+          return WalkResult::interrupt();
+        }
+        return WalkResult::advance();
+      });
+  if (!scfLoopOp) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "There is no scf.for/forall loop to fuse with\n");
+    return;
+  }
 
-    Operation *terminatorStoreOp = *(users.begin());
-    if (!(isa<tensor::InsertSliceOp, tensor::ParallelInsertSliceOp>(
-            terminatorStoreOp))) {
-      terminatorStoreOp->emitOpError(
-          "Expected either tensor.insert_slice OR tensor.parallel_insert_slice "
-          "to be the only user of the compute op");
-      return signalPassFailure();
-    }
+  // Step 3. Search the compute op and its consumer slices.
+  linalg::LinalgOp linalgOp;
+  scfLoopOp->walk<WalkOrder::PostOrder, ReverseIterator>(
+      [&](linalg::LinalgOp op) {
+        linalgOp = op;
+        return WalkResult::interrupt();
+      });
+  if (!linalgOp) {
+    LLVM_DEBUG(llvm::dbgs() << "Could not find any compute op\n");
+    return;
+  }
 
-    std::optional<scf::SCFFuseConsumerOfSliceResult> fusedConsumer =
-        scf::tileAndFuseConsumerOfSlice(rewriter, terminatorStoreOp);
-    if (!fusedConsumer) {
-      terminatorStoreOp->emitOpError(
-          "Failed to fuse any consumer op into the producer");
-      return signalPassFailure();
-    }
-    fusedConsumer->origConsumerOperand->getOwner()->erase();
+  Operation *computeOp = linalgOp;
+  // Step 4. Based on the `fuseDepth`, we would greedily fuse the consumer ops.
+  for (unsigned depth = 1; depth <= fuseDepth; depth++) {
+    do {
+      Value::user_range users = computeOp->getResult(0).getUsers();
+      if (!llvm::hasSingleElement(users)) {
+        computeOp->emitOpError("Expected only one user of the compute op");
+        return signalPassFailure();
+      }
+
+      Operation *terminatorStoreOp = *(users.begin());
```

Comment:
Why is this called `teminatorStoreOp`? What about just `userOp`?

Reply:
The upstream consumer fusion API `scf::tileAndFuseConsumerOfSlice` takes in either `tensor.insert_slice` or `tensor.parallel_insert_slice` - both of these are actually expected to be the final operation in the `scf.for*` op (it in turn verifies these claims).

So, the naming of this variable as `teminatorStoreOp` is a preferential choice - I believe it helps to know what is the nature of this user op.

Reply:
Are `tensor::InsertSliceOp, tensor::ParallelInsertSliceOp` terminator store operations? The `scf:;tileAndFuseConsumerOfSlice` function calls them `candidateSliceOp` instead. I am nitting here but I don't find `terminatorStoreOp` a clear name while reading the code and I couldn't find it anywhere else in the code base. What about `candidateSliceOp` like the scf API?